[RF-Docs] [ci-skip]  Active Storage Guide (#56340)
rails/rails: guides/source/active_storage_overview.md

After reading this guide, you will know:

* How to attach one or many files to a record.
* How to display attached files and how to delete them.
* How to use variants to transform images.
* How to generate an image representation of a non-image file (e.g. PDF).
* How to send file uploads directly from browsers to a storage service.
* How to setup cloud storage services to work with Active Storage.

--------------------------------------------------------------------------------

What is Active Storage?
-----------------------

Active Storage facilitates attaching files to Active Record objects and

> uploading those files to cloud storage services.

Reviewer: Is it worth rewording this?

```suggestion
uploading those files to your server or to a cloud storage service.
```

* How to generate an image representation of a non-image file (e.g. PDF).
* How to send file uploads directly from browsers to a storage service.
* How to setup cloud storage services to work with Active Storage.

--------------------------------------------------------------------------------

What is Active Storage?
-----------------------

Active Storage facilitates attaching files to Active Record objects and
uploading those files to your server or to a cloud storage service.

Active Storage supports image variants (e.g. resizing) and can transform and
store variants of uploaded images. Using Active Storage, you can also generate
image representations of non-image uploads like PDFs and videos, and extract
metadata.

For cloud storage services, Active Storage supports mirroring files to secondary
services to serve as a backup or to allow migration between services. Active
Storage also supports Direct Uploads, allowing files to be uploaded straight
from the client's browser to the configured cloud storage service. This avoids
routing large files through your Rails servers.

Active Storage also supports a `Disk` service which uses the local filesystem by default.

> ### Setup

Reviewer 2: In other guides "Setup" or "Installation" is a top section.
- [Action Mailbox > Setup](https://edgeguides.rubyonrails.org/action_mailbox_basics.html#setup)
- [Action Text > Installation](https://edgeguides.rubyonrails.org/action_text_overview.html#installation)
- [Active Storage > Setup](https://edgeguides.rubyonrails.org/active_storage_overview.html#setup)
- [I18n > Setup the Rails Application for Internationalization](https://edgeguides.rubyonrails.org/i18n.html#setup-the-rails-application-for-internationalization)

For consistency we should probably do it here as well?
```suggestion
Setup
-----
```

Author: sure!

```

The services configured in the `config/storage.yml` file are then used in
environment specific configuration files. For example, in order to use the
`local` service above during development, we modify the
`config/environments/development.rb` file:

```ruby
# config/environments/development.rb
config.active_storage.service = :local
```

The `config/storage.yml` file is also where cloud services can be configured.
For example, assuming there is a serviced called `amazon` in the
`config/storage.yml` file, in order to use that service in production:

```yml
# config/environments/production.rb
config.active_storage.service = :amazon
```

You can find detailed information about [configuring cloud
services](#configuring-cloud-services) in a later section.

### Configuring Active Storage Routes

Active Storage automatically adds routes to your application for serving files.
These routes are mounted under `/rails/active_storage` by default. For example,
So when someone requests a file attachment in your app, the URL may look like
`https://example.com/rails/active_storage/blobs/redirect/eyJf.../photo.jpg`. You can see all the routes by running:

```bash

> bin/rails routes | grep active_storage

Reviewer 2: This is missing the prompt. Also we should use the build-in grep option as that keeps the table headers in the output.
```suggestion
$ bin/rails routes --grep active_storage
```

### `has_many_attached`

The [`has_many_attached`][] method sets up a one-to-many relationship between a
record and attached files. Each record can have many files attached to it.

For example, suppose your application has a `Product` model. Each product can have multiple images associated with it:

```ruby
class Product < ApplicationRecord
  has_many_attached :images
end
```

You can also use the model generator command like this:

```bash
$ bin/rails generate model Product images:attachments
```

The controller to create a product with multiple images looks like this:

```ruby
class ProductsController < ApplicationController
  def create
    product = Product.create!(product_params)
    redirect_to product
  end

  private
    def product_params
      params.expect(product: [ :title, :content, images: [] ])
    end
end
```

You can call [`images.attach`][Attached::Many#attach] to add new images to an existing product:

```ruby
@product.images.attach(params[:images])
```

You can call [`images.attached?`][Attached::Many#attached?] to determine whether a particular product has any images:

```ruby
@product.images.attached?
```

> NOTE: When using `has_many_attached`, calling `images.attach(...)` adds new attachments to the list of existing attachments. It does not replace or overwrite existing attachments. If you want to replace existing images, you must explicitly [purge](#removing-files) the old attachments before attaching new ones.

Author: No worries at all @p8! This has been a bit confusing so glad we are clarifying it in the guide now! 

Oh and I like that gist, nice minimal way to test out a feature. I keep spinning up demo Rails apps, haha.

example, the following code retrieves all `Product` records whose attached
images are videos with an MP4 format:

```ruby
class Product < ApplicationRecord
  has_many_attached :images
end

products = Product.joins(:images_blobs).where(
  active_storage_blobs: { content_type: "video/mp4" }
)
```

This query executes against the `active_storage_blobs` table rather than the
attachment records themselves, since the join created by `joins(:images_blobs)`
operates on the blob side of the association. You can combine such blob-based
filters with additional scope conditions in the same way you would with any
standard Active Record query.

Serving Files
-------------

Active Storage can serve files in two different ways: redirect mode and proxy
mode. Both modes use built-in controllers to deliver blobs and
[representations](#file-representations), but they differ in how the file
ultimately reaches the browser.

WARNING: All Active Storage controllers are publicly accessible by default. The
Rails generated URLs are hard to guess, but permanent by design. If your files
require a higher level of protection consider implementing [Authenticated
Controllers](#authenticated-controllers).

> ### Redirect Mode

Reviewer 2: @bhumi1102 I recently had a look at these controllers for #55865 and found out you can also create expiring URLs:
https://edgeapi.rubyonrails.org/classes/ActiveStorage/Blobs/RedirectController.html
https://edgeapi.rubyonrails.org/classes/ActiveStorage/Blobs/ProxyController.html

I think we should mention the `:expires_at` option and `config.active_storage.urls_expire_in` in this guide as well.

Author: Hi @p8 - agree, thanks for the heads up on this PR! I read the details and the blog post linked. I've added a little subsection to explain the `expires_in` option. Hopefully this conveys the difference between access control with authentication controller and other options. Let me know if you have notes on that new section and I'll address it to get this one guide ready to go!

)
```

### `has_many_attached`

Similarly, when you use `has_many_attached`, Rails defines two associations: a
`has_many` association named `<name>_attachments`, which represents the join
records in the `active_storage_attachments` table, and a `has_many :through`
association named `<name>_blobs`, which gives access to the corresponding rows
in `active_storage_blobs` table.

Because the `_blobs` association provides a normal relational join, you can
query it directly to filter records based on metadata stored in the blob. For
example, the following code retrieves all `Product` records whose attached
images are videos with an MP4 format:

```ruby
class Product < ApplicationRecord
  has_many_attached :images
end

products = Product.joins(:images_blobs).where(
  active_storage_blobs: { content_type: "video/mp4" }
)
```

This query executes against the `active_storage_blobs` table rather than the
attachment records themselves, since the join created by `joins(:images_blobs)`
operates on the blob side of the association. You can combine such blob-based
filters with additional scope conditions in the same way you would with any
standard Active Record query.

> Serving Files

Reviewer 2: This was a pleasure to read.

Author: Oh good, thank you :)

There are several ways to include the Active Storage JavaScript library in your application:

1. Directly via `javascript_include_tag` - include the library in your HTML without bundling through the asset pipeline. Autostart is enabled automatically:

```erb
<%= javascript_include_tag "activestorage" %>
```

1. Using Importmap (ESM) - pin the library in `config/importmap.rb`:

```ruby
pin "@rails/activestorage", to: "activestorage.esm.js"
```

Then import and start it in your HTML:

```html
<script type="module-shim">
  import * as ActiveStorage from "@rails/activestorage"
  ActiveStorage.start()
</script>
```

1. Using the Asset Pipeline - require the library in a JavaScript manifest:

```javascript
//= require activestorage
```

1. Using the npm package - install via npm/yarn and import it in your JavaScript bundle:

```js
import * as ActiveStorage from "@rails/activestorage"
ActiveStorage.start()
```

All of these approaches provide the same functionality; choose the one that matches your application’s JavaScript setup.

> Next step is to set the `direct_upload: true` option in your [`file_field`

Reviewer 2: Wondering if it would be clearer to add a heading here as well?

```suggestion
#### Enabling Direct Uploads on the Input

The next step is to set the `direct_upload: true` option in your [`file_field`
```

Sometimes you need to process a file after it’s uploaded. For example, to
convert it to a different format. You can use the [`download`][Blob#download]
method to read the file's binary data (i.e. blob) into memory:

```ruby
binary = user.profile_photo.download
```

You can also download a file's blob to local disk so an external program (e.g. a
virus scanner or media transcoder) can operate on it. In the example below, the
blob's [`open`][Blob#open] method saves the file to a tempfile on disk and then
yields the file to the block:

```ruby
product.images.open do |file|
  system "/path/to/virus/scanner", file.path
  # ...
end
```

NOTE: Active Storage attachments are not fully available until the record’s
transaction has committed. This means methods like `download` and `open` cannot
be used reliably inside an `after_create` callback because the blob is not
persisted yet. Use `after_create_commit` if you need to process the uploaded
file immediately after creation.

[Blob#download]:
https://api.rubyonrails.org/classes/ActiveStorage/Blob.html#method-i-download
[Blob#open]:
https://api.rubyonrails.org/classes/ActiveStorage/Blob.html#method-i-open

Removing Files
--------------

> 

Reviewer 2: It would be nice if we added some introduction paragraph here explaining when you want to remove files.

There also is the `detach` method, that we don't describe.
It removes the attachment and leaves the Blob in place.
https://edgeapi.rubyonrails.org/classes/ActiveStorage/Attached/One.html#method-i-detach

I'm not sure what the use-case is (making sure a Blob isn't accidentally deleted?) but we probably need to add it?

Author: Good points. Added an intro sentence. Also added `detach` method (in the 2nd section along with `unattached`). It seems the use case may be when you want to leave the file on the storage service but not have it associated with a record? Either way, I added a description with an example.

> ### Purging Attachments From a Model

Reviewer 2: I noticed there is a `purge` method defined for `has_many_attached` as well. It purges all attached files.
Should we describe that method as well?

Or maybe more importantly explain how to remove a single file from a `has_many_attached`?

Also if we'll add the `detach` method, maybe rename this section to "Removing Attachments from a Model"?

Author: Agree. Renamed. Added an example to show how to remove a single file from `has_many_attached`

To remove an attachment from a model, call [`purge`][Attached::One#purge] on the
attachment. If your application is set up to use Active Job, removal can be done
in the background instead by calling [`purge_later`][Attached::One#purge_later].

> Purging deletes the blob and the file from the storage service.

Reviewer 2: Nitpick. Some readers might think the blob is deleted from the storage service as well.

Also attachments are deleted/destroyed as well and blobs are only destroyed when there are no more attachments attached (I don't think we mention anywhere in the guide that only one Blob is created for a file with the same checksum).

Maybe something like:
```suggestion
Purging destroys the attachment record. If the blob has no more attachments it gets destroyed as well and the file is deleted from the storage service.
```

Author: good point!

### Purging Unattached Uploads

There are cases where a file is uploaded but never attached to a record. This
can happen when using [Direct Uploads](#direct-uploads). You can query for
unattached records using the [unattached scope][Blob.unattached]. Below is an
example using a [custom rake task](command_line.html#custom-rake-tasks).

[Blob.unattached]:
https://api.rubyonrails.org/classes/ActiveStorage/Blob.html#method-c-unattached

```ruby
namespace :active_storage do
  desc "Purges unattached Active Storage blobs. Run regularly."
  task purge_unattached: :environment do
    ActiveStorage::Blob.unattached.where(created_at: ..2.days.ago).find_each(&:purge_later)
  end
end
```

WARNING: The query generated by `ActiveStorage::Blob.unattached` can be slow and
potentially disruptive on applications with larger databases.

Analyzing Files For Metadata
----------------------------

Active Storage analyzes files to extract metadata like image dimensions, video
duration, and audio bit rate. Analyzed files will store additional information
in the metadata hash, including `analyzed: true`. You can check whether a blob

> has been analyzed by calling the [`analyzed?`][] method on it.

Reviewer 2: Before describing `analyzed?` I think we should describe the `metadata` method (I've just added this method to the API docs).
Something like:
````suggestion
Active Storage analyzes files to extract metadata like image dimensions, video
duration, and audio bit rate.  Once a file has been analyzed, the metadata is stored in the
`active_storage_blobs` table and can be viewed with
the [`metadata`][] method :

```irb
> user.profile_photo.metadata
=> {"identified" => true, "width" => 112, "height" => 243, "created_at" => "2026-04-05T00:11:48+02:00", "analyzed" => true}
```

Analyzed files will store additional information
in the metadata hash, including `analyzed: true`. You can check whether a blob
has been analyzed by calling the [`analyzed?`][] method on it.
````

Author: Thank you for that addition and the diff for it!

not have to pass through your Rails server.

Direct uploads integrate seamlessly with Active Storage’s attachments and
variants, allowing you to use the same models, validations, and background
processing workflows as standard uploads.

Active Storage, with its included JavaScript library, supports uploading
directly from the client to the cloud.

### Setup JavaScript Library

In order to start using direct uploads, you'll need to use the JavaScript
Library included with Active Storage. The library handles:

* Initiating uploads to the configured service (e.g., S3, GCS, Azure).
* Tracking upload progress and reporting it to the user.
* Updating form inputs with the necessary signed IDs so that Rails can associate
  the uploaded file with the model when the form is submitted.

To use direct uploads, you'll need to include the library in your application’s
JavaScript bundle and enable the `direct_upload: true` option on your file input
fields. This allows Rails and the storage service to coordinate securely using
signed IDs, without requiring extra backend configuration.

There are several ways to include the Active Storage JavaScript library in your

> application:

Reviewer 2: This list isn't properly displaying:

<img width="857" height="694" alt="Image" src="https://github.com/user-attachments/assets/9919c9e0-fc6b-436d-8bfc-9d73155bec32" />

Author: man, I thought I fixed this. alright, I might need to just go with subheading like you suggested a while back haha

  end
end
```

WARNING: The query generated by `ActiveStorage::Blob.unattached` can be slow and
potentially disruptive on applications with larger databases.

There is also a [`detach`](https://edgeapi.rubyonrails.org/classes/ActiveStorage/Attached/One.html#method-i-detach) method, which deletes the associated attachments but leaves the blobs in place. This intentionally orphans the blob and leaves the file on the storage service.

```ruby
user.profile_photo.detach
```

This can be useful if you want to disassociate a file from a record without deleting it from storage, in case the blob is referenced elsewhere. Note that you can later find such orphaned blobs using the `unattached` scope if needed.

Analyzing Files For Metadata
----------------------------

Active Storage analyzes files to extract metadata like image dimensions, video
duration, and audio bit rate.  Once a file has been analyzed, the metadata is stored in the
`active_storage_blobs` table and can be viewed with
the [`metadata`][] method :

```irb
> user.profile_photo.analyzed?
=> true
```

Once a file has been analyzed, the metadata is stored in the
`active_storage_blobs` table and can be viewed with

> `user.profile_photo.blob.metadata`.

Reviewer 2: I think this suggestion didn't apply correctly.

````suggestion
```irb
> user.profile_photo.metadata
=> {"identified" => true, "width" => 112, "height" => 243, "created_at" => "2026-04-05T00:11:48+02:00", "analyzed" => true}
```
````

I don't have a strong opinion about keeping the `analyzed?` method example.
So it could also be:

````suggestion
```irb
> user.profile_photo.metadata
=> {"identified" => true, "width" => 112, "height" => 243, "created_at" => "2026-04-05T00:11:48+02:00", "analyzed" => true}
```

You can check whether a blob has been analyzed by calling the [`analyzed?`][]
method on it:

```irb
> user.profile_photo.analyzed?
=> true
```
````

You can set the application level default for the `analyze` option in your Rails
application configuration as well:

```ruby
# config/application.rb
config.active_storage.analyze = :later
```

### Validating Attachment Metadata

Since attachments are analyzed immediately by default, metadata is available for
model validations. For example, it's possible to validate that the uploaded
profile photo has certain dimensions:

```ruby
class User < ApplicationRecord
  has_one_attached :profile_photo

  validate :validate_profile_photo_size, if: -> { profile_photo.attached? }

  private
    def validate_profile_photo_size
      if profile_photo.metadata[:width] < 200 || profile_photo.metadata[:height] < 200
        errors.add(:profile_photo, "must be at least 200x200 pixels")
      end
    end
end
```

NOTE: Since [Direct uploads](#direct-uploads) bypass the server, files aren't
locally available for analysis. In this case, `:immediately` falls back to
`:later`, analyzing via background job after upload completes. So model
validations using metadata isn't possible. You can validate on the client side
using JavaScript instead.

> [`analyzed?`]:
> https://api.rubyonrails.org/classes/ActiveStorage/Blob/Analyzable.html#method-i-analyzed-3F

Reviewer 2: ```suggestion
[`metadata`][]:
https://api.rubyonrails.org/classes/ActiveStorage/Blob.html#method-i-metadata
[`analyzed?`]:
https://api.rubyonrails.org/classes/ActiveStorage/Blob/Analyzable.html#method-i-analyzed-3F
```

NOTE: Using [Direct Uploads](#direct-uploads) can sometimes result in a file that uploads, but never attaches to a record. Consider [purging unattached uploads](#purging-unattached-uploads).

> Testing

Reviewer 2: For some other guides we link to the testing guide instead.
For example:
https://guides.rubyonrails.org/active_job_basics.html#job-testing
https://guides.rubyonrails.org/action_cable_overview.html#testing

Maybe move this as well?
